File: libs/power-log-parser/src/lib/obfuscator.ts
import { BlockType, CardIds, GameTag } from '@firestone-hs/reference-data';
import { Action } from './models/action';
import { FullEntity } from './models/entity';
import { Node, NodeType } from './models/node';
import type { GameState } from './state/game-state';

export class Obfuscator {
	static shouldObfuscateCardDraw(
		entity: FullEntity,
		gameState: GameState,
		node: Node,
		isPlayer: boolean,
		revealed = false,
	): boolean {
		const debug = entity.Entity === 158;
		if (node?.Parent?.Type === NodeType.Action) {
			const action = node.Parent.Object as Action;
			const actionEntityId = action.Entity;
			const actionEntity = gameState.CurrentEntities.get(actionEntityId);
			if (action.Type === (BlockType.POWER as number)) {
				const actionCardId = actionEntity?.CardId;
				switch (actionCardId) {
					case CardIds.SirFinleySeaGuide:
						return true;
				}
			}
		}

		// This is here to prevent info leaks.
		// However, it overrides any other logic we can have - for instance, mimicry should let you
		// know what is what
		if (
			!isPlayer &&
			entity.AllPreviousTags.find((t) => t.Name === (GameTag.IS_USING_TRADE_OPTION as number) && t.Value === 1) !=
				null
		) {
			return true;
		}

		if (
			!isPlayer &&
			// Always obfuscate for the opponent if the card is not marked as being revealed
			// For instance, Deathwing creates cards in the opponent's deck, and they are barebones, without any
			// additional tag. So Hidden is false in that case, as well as revealed.
			// UPDATE 2026-06-18: not needed right now, but I think we'll have to comment it at some point
			entity.Hidden &&
			!revealed &&
			entity.GetTag(GameTag.CASTS_WHEN_DRAWN) !== 1 &&
			entity.GetTag(GameTag.SUMMONED_WHEN_DRAWN) !== 1
		) {
			return true;
		}

		return false;
	}

	/**
	 * When card identity is obfuscated, still pass creator provenance if the entity came from deck.
	 * Game-state gates what is safe to use ({@link publicCardGiftCreators}, isCreatorPublic, etc.).
	 */
	static creatorCardIdForDrawEvent(
		shouldObfuscate: boolean,
		wasInDeck: boolean,
		creatorCardId: string | null | undefined,
	): string | null {
		if (!creatorCardId) {
			return null;
		}
		if (!shouldObfuscate) {
			return creatorCardId;
		}
		return wasInDeck ? creatorCardId : null;
	}
}

File: tools/linux-helpers/hs-converter/HearthstoneReplays/Events/Obfuscator.cs
using HearthstoneReplays.Enums;
using HearthstoneReplays.Parser;
using HearthstoneReplays.Parser.ReplayData.Entities;
using static HearthstoneReplays.Events.CardIds;
using Action = HearthstoneReplays.Parser.ReplayData.GameActions.Action;

namespace HearthstoneReplays.Events
{
    internal class Obfuscator
    {
        internal static bool shouldObfuscateCardDraw(FullEntity entity, GameState gameState, Node node, bool isPlayer, bool revealed = false)
        {
            var updatedEntity = gameState.CurrentEntities.GetValueOrDefault(entity.Id);
            if (node?.Parent?.Type == typeof(Action))
            {
                var action = node.Parent.Object as Action;
                var actionEntityId = action.Entity;
                var actionEntity = gameState.CurrentEntities.GetValueOrDefault(actionEntityId);
                if (action.Type == (int)BlockType.POWER)
                {
                    var actionCardId = actionEntity?.CardId;
                    switch (actionCardId)
                    {
                        case SirFinleySeaGuide:
                            return true;
                    }
                }
            }

            // Avoid info leaks for cards that have been traded into the deck. This is necessary because they
            // are logged in clear in the Power.log
            if (!isPlayer && entity.AllPreviousTags.Find(t => t.Name == (int)GameTag.IS_USING_TRADE_OPTION && t.Value == 1) != null) {
                return true;
            }
            // What can happen is the card is revealed when drawn (eg Temporal Anomaly), then immediately hidden again
            if (!isPlayer && entity.Hidden && !revealed && (entity.GetTag(GameTag.CASTS_WHEN_DRAWN) != 1 && entity.GetTag(GameTag.SUMMONED_WHEN_DRAWN) != 1))
            {
                return true;
            }

            return false;
        }
    }
}
